File: assignment_6.go
package main

import (
	"fmt"
	"math/rand"
	"sync"
	"time"
)

// Task represents a unit of work
type Task struct {
	ID int
}

// process simulates computation
func (t Task) process() error {
	fmt.Printf("%s is processing Task %d\n", getGoroutineName(), t.ID)
	time.Sleep(time.Millisecond * time.Duration(300+rand.Intn(400))) // simulate delay
	return nil
}

// getGoroutineName returns a pseudo-name for logging (since Go doesn't expose goroutine ID)
func getGoroutineName() string {
	return fmt.Sprintf("Worker-%d", rand.Intn(1000)) // pseudo-ID
}

// worker function pulls from task channel and processes tasks
func worker(id int, tasks <-chan Task, results *[]string, mu *sync.Mutex, wg *sync.WaitGroup) {
	defer wg.Done()
	name := fmt.Sprintf("Worker-%d", id)

	for task := range tasks {
		err := task.process()
		if err != nil {
			fmt.Printf("%s encountered an error: %v\n", name, err)
			continue
		}
		// Save result with locking
		mu.Lock()
		*results = append(*results, fmt.Sprintf("%s processed Task %d", name, task.ID))
		mu.Unlock()
	}
	fmt.Println(name + " finished processing.")
}

func main() {
	rand.Seed(time.Now().UnixNano())

	// Create shared task channel and results list
	taskChan := make(chan Task, 20)
	var results []string
	var mu sync.Mutex
	var wg sync.WaitGroup

	// Fill the task channel
	for i := 1; i <= 20; i++ {
		taskChan <- Task{ID: i}
	}
	close(taskChan) // No more tasks will be sent

	// Launch worker goroutines
	numWorkers := 4
	for i := 1; i <= numWorkers; i++ {
		wg.Add(1)
		go worker(i, taskChan, &results, &mu, &wg)
	}

	// Wait for all workers to finish
	wg.Wait()

	// Print results
	fmt.Println("\nFinal Results:")
	for _, res := range results {
		fmt.Println(res)
	}
}

File: assignment_6.java
import java.util.*;

public class DataProcessingSystem {

    // Represents a unit of work
    static class Task {
        private final int id;

        public Task(int id) {
            this.id = id;
        }

        public void process() throws InterruptedException {
            System.out.println(Thread.currentThread().getName() + " is processing Task " + id);
            Thread.sleep(500); // Simulate processing delay
        }

        public int getId() {
            return id;
        }
    }

    // Shared queue with synchronized access
    static class TaskQueue {
        private final Queue<Task> queue = new LinkedList<>();

        public synchronized void addTask(Task task) {
            queue.add(task);
            notify(); // Wake up one waiting thread if any
        }

        public synchronized Task getTask() throws InterruptedException {
            while (queue.isEmpty()) {
                wait(); // Wait until a task is available
            }
            return queue.poll();
        }

        public synchronized boolean isEmpty() {
            return queue.isEmpty();
        }
    }

    // Worker thread that pulls and processes tasks
    static class Worker extends Thread {
        private final TaskQueue taskQueue;
        private final List<String> results;

        public Worker(TaskQueue taskQueue, List<String> results, String name) {
            super(name);
            this.taskQueue = taskQueue;
            this.results = results;
        }

        @Override
        public void run() {
            while (true) {
                Task task;
                try {
                    synchronized (taskQueue) {
                        if (taskQueue.isEmpty()) break;
                    }

                    task = taskQueue.getTask(); // May block if queue is empty
                    task.process();

                    synchronized (results) {
                        results.add(getName() + " processed Task " + task.getId());
                    }

                } catch (InterruptedException e) {
                    System.err.println(getName() + " was interrupted.");
                    break;
                } catch (Exception e) {
                    System.err.println(getName() + " encountered an error: " + e.getMessage());
                }
            }

            System.out.println(getName() + " finished processing.");
        }
    }

    // Main entry point
    public static void main(String[] args) {
        TaskQueue taskQueue = new TaskQueue();
        List<String> results = Collections.synchronizedList(new ArrayList<>());

        // Add some tasks
        for (int i = 1; i <= 20; i++) {
            taskQueue.addTask(new Task(i));
        }

        // Create and start worker threads
        int numWorkers = 4;
        List<Worker> workers = new ArrayList<>();
        for (int i = 0; i < numWorkers; i++) {
            Worker worker = new Worker(taskQueue, results, "Worker-" + (i + 1));
            workers.add(worker);
            worker.start();
        }

        // Wait for all workers to finish
        for (Worker worker : workers) {
            try {
                worker.join();
            } catch (InterruptedException e) {
                System.err.println("Main thread interrupted.");
            }
        }

        // Show results
        System.out.println("\nFinal Results:");
        for (String result : results) {
            System.out.println(result);
        }
    }
}
